components/style/context.rs

```diff
@@ -76,25 +75,25 @@ pub struct SharedStyleContext {
     pub quirks_mode: QuirksMode,
 }
 
-pub struct ThreadLocalStyleContext {
-    pub style_sharing_candidate_cache: RefCell<StyleSharingCandidateCache>,
+pub struct ThreadLocalStyleContext<E: TElement> {
+    pub style_sharing_candidate_cache: StyleSharingCandidateCache<E>,
```

Comment:
I think we should annotate the lifetime of the cache with the lifetime of the scope. I don't think it'd be too hard to propagate that lifetime, what do you think? Seems better to have the scope lifetime there and in `DomTraversalContext` than in `top_down_dom`.